man/man1/ccdprint.1
.TH CCDPRINT 1NEMO "26 January 2021"

.SH "NAME"
ccdprint \- print out map values

.SH "SYNOPSIS"
.PP
\fBccdprint in=\fP\fIimage\fP [parameter=value]

.SH "DESCRIPTION"
\fBccdprint\fP prints out map values in an image. The grid is
specified in x, y and z independantly, and can be any
integer array, including implied do loops, e.g.
\fIx=0,10,40:50:2,90,100\fP will print out values at 
\fIx=0,10,40,42,44,46,48,50,90,100\fP. Note that the first grid
point is 0 by convention in NEMO, though this program can use
\fBoffset=1\fP to "think" like a fortran array if that is
convenient.
.PP
By default the matrix is printed such
that the bottom left corner of the output is the first pixel,
as in (astronomical) image display. However, by setting
\fByreverse=f\fP it can be printed with the first pixel on the
top right. This effect can also be achieved by an appropriate
choice of \fBy=NY-1:1:-1\fP, but the user is then required
to know \fBNY\fP!
.PP
The physical coordinates can be printed out optionally, where the output
format is controlled by the \fBnewline=\fP parameter. See also
\fIxy2sky(1)\fP from \fiwcstools(5NEMO)\fP.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
input file, in \fIimage(5NEMO)\fP format [no default].
.TP
\fBx=\fIx1,x2,...\fP
Grid coordinates in X to display. 
If no values given, all gridpoints along this axis will be selected.
Note the first gridpoint is 0. By default this program will thus
display the first pixel value.
[Default: \fB0\fP].
.TP
\fBy=\fIy1,y2,...\fP
Grid coordinates in Y to display. See \fBx=\fP.
[Default: \fB0\fP].
.TP
\fBz=\fIz1,z2,...\fP
Grid coordinates in Z to display. See \fBx=\fP.
[Default: \fB0\fP].
.TP
\fBscale=\fIscale_factor\fP
scale factor with which the data is multiplied before being displayed
[default: \fB1\fP].
.TP
\fBformat=\fIfmt-expr\fP
Real-valued C-format expression (see also printf(3), 
e.g. %10.6e, %5.2f) with which data and optional
coordinate values are displayed
[default: \fB%g\fP].
.TP
\fBnewline=t|f\fP
Add a newline between each map value
printed? If false, all rows are printed
on one line, and a blank line separates different planes.
[Default: \fBf\fP]
.TP
\fBlabel=\fP\fIx,y,z\fP
Any combination of \fBx\fP, \fBy\fP and/or \fBz\fP can be given, which labels
the output with their X, Y and/or Z physical values. 
If \fBnewline\fP has been
set to true, the selected labels are printed for each coordinate selected
(this produces a nicely formatted table that can be easily used by
subsequent analysis programs), whereas the default value would
print a row of X coordinates before each printed plane , and a left
column of Y coordinates, if selected. Each new plane is preceded with
a plane coordinate also. This option is more readable for humans, but
not easily processed by other programs.
[Default: no labels].
.TP
\fBoffset=0|1\fP
Should X,Y,Z be 0- or 1- based. NEMO images use 0-based images,
but to appease fortran users, 1-based input can be given here
if this parameter is set to 1. Default: 0
.TP
\fByreverse=t|f\fP
Reverse printing the Y values. By default this is true, such that the
printed layout resembles how it is normally displayed in an
image display (at least astronomically). 
[Default: \fBt\fP]
.TP
\fBpixel=t|f\fP
Print out X,Y,Z labels in physical or pixel coordinates. By default
physical. [Default: \fBf\fP]
.TP
\fBpair=t|f\fP
Should input (x,y,z) be paired up [f]
.TP
\fBseq=\fP
Print a sequence using access shortcut [0]

.SH "SEE ALSO"
xy2sky(1), sky2xy(1), printf(3), wcstools(5NEMO), image(5NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
~src/image/io	ccdprint.c ccdprint.1
.fi

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
23-jan-89	V1.0: Created             	PJT
8-jul-93	V1.1: added defaults 'all' for blank x=, y=, z=  	PJT
9-jul-93	V1.2: added label=	PJT
28-jul-02	V1.3: documented offset=, added pixel=	PJT
8-nov-05	V1.4: added yreverse= and better handling of blank lines	PJT
26-jan-2021	V1.7: use reference pixel	PJT
.fi


src/image/io/ccdprint.c
/* 
 * CCDPRINT: print values at gridpoints, with optional numeric labels
 *
 *	 23-jan-89  V1.0 created        Peter Teuben
 *	 15-jul-89  V1.0a   - nemoinpX instead of nemoinp	PJT
 *	  8-jul-93  V1.1 blank default for x=,y=,z=             pjt
 *		    and  fixep various old style coding
 *	 14-sep-93  V1.2 Added index mode keyword offset=	pjt
 *	 14-oct-99  V1.2b fixed bug printing WCS labels		pjt
 *       28-jul-02  V1.3  allow coordinates printed as pixel (0...N-1) pjt
 *        8-nov-05  V1.4  cleanup for prototypes, better blank line handling  pjt
 *                        also added yreverse=
 *       24-jan-06  V1.5  pairing allowed, but crummy
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <image.h>

string defv[] = {
  "in=???\n           Input filename",
  "x=0\n              Pixels in X to print (0=1st pixel, leave blank for all)",
  "y=0\n              Pixels in Y to print",
  "z=0\n              Pixels in Z to print",
  "scale=1.0\n        Scale factor for printout",
  "format=%g\n        Format specification for output",
  "newline=f\n        Force newline between each number?",
  "label=\n	      Add x, y and or z labels add appropriate labels",
  "offset=0\n         Offset (0 or 1) to index coordinates X,Y,Z",
  "yreverse=t\n       Reverse printing Y values?",
  "pixel=f\n          Labels in Pixel or Physical coordinates?",
  "pair=f\n           Should input (x,y,z) be paired up",
  "seq=0\n            Print a sequence using access shortcut",
  "VERSION=1.7\n      26-jan-2021 PJT",
  NULL,
};

string usage = "print values at gridpoints of an image";

string cvsid="$Id$";

int ini_array(string key, int *dat, int ndat, int offset);
void myprintf(string fmt, real v);


void nemo_main(void)
{
    int     *ix, *iy, *iz, i, j, k, j1, l, nx, ny, nz, nxpos, nypos, nzpos, offset, nlcount=0;
    int     nmax;
    bool    newline, newline1, newline2, newline3, xlabel, ylabel, zlabel, Qpixel, Qyrev, Qpair;
    string  infile;			        /* file name */
    stream  instr;				/* file stream */
    imageptr iptr=NULL;			      /* allocated dynamically */
    string   fmt, label;
    real     scale_factor, x, y, z, f, *data;
    int      seq;

    scale_factor = getdparam("scale");
    fmt = getparam("format");
    instr = stropen (getparam("in"), "r");
    newline = getbparam("newline");
    Qpixel = getbparam("pixel");
    Qyrev = getbparam("yreverse");
    Qpair = getbparam("pair");
    label = getparam("label");
    xlabel = scanopt(label,"x");
    ylabel = scanopt(label,"y");
    zlabel = scanopt(label,"z");
    offset = getiparam("offset");
    seq = getiparam("seq");
    if (read_image (instr,&iptr) == 0)
      error("Problem reading image from in=",getparam("in"));

    if (seq) {
      data = Frame(iptr);
      for (i=0; i<seq; i++) printf("%g\n",data[i]);
      stop(0);
    }
    
    nx = Nx(iptr);	                        /* cube dimensions */
    ny = Ny(iptr);
    nz = Nz(iptr);
    ix = (int *) allocate(nx*sizeof(int));        /* allocate position arrays */
    iy = (int *) allocate(ny*sizeof(int));
    iz = (int *) allocate(nz*sizeof(int));
    nxpos = ini_array("x",ix,nx,offset);
    nypos = ini_array("y",iy,ny,offset);
    nzpos = ini_array("z",iz,nz,offset);
    if (Qpair) {
      warning("new pairing mode, not all options allowed");
      nmax = MAX(nxpos,nypos);
      nmax = MAX(nmax,nzpos);
      for (l=0; l<nmax; l++) {
	i = ix[nxpos>1 ? l : 0];
	j = iy[nypos>1 ? l : 0];
	k = iz[nzpos>1 ? l : 0];
	f = CubeValue(iptr,i,j,k);
	printf (fmt,f*scale_factor);	  
	printf(" ");
	if (newline)
	  printf("\n");
      }
    }

    newline1 = newline2 = newline3 = newline;
    if (!hasvalue("x") && !hasvalue("y")){
      if ((nz>1 && !hasvalue("z")) || nz<=1) {
	newline1 = TRUE;
	newline2 = FALSE;
      }
    }

    for (k=0; k<nzpos; k++) {				/* loop over all planes */
        if (Qpair) break;
        z = Zmin(iptr) + (iz[k]-Zref(iptr)) * Dz(iptr);
        if (!newline && zlabel) {
            printf("plane Z = ");
	    myprintf(fmt, Qpixel ? (real)k : z);
            printf("\n\n");
	    nlcount += 2;
        }
        for (j1=0; j1<nypos; j1++) {			/* loop over all rows */
	    if (Qpair && nypos > 1) j1 = l;
	    j = Qyrev ? nypos-1-j1 : j1;
            y = Ymin(iptr) + (iy[j]-Yref(iptr)) * Dy(iptr);
            if (j==(nypos-1) && !newline && xlabel) {    /* pr. row of X coords */
	      if (ylabel) printf(" Y\\X ");     /* ? how to get correct length ? */
	      for (i=0; i<nxpos; i++) {
		x = Xmin(iptr) + (ix[i]-Xref(iptr)) * Dx(iptr);
		myprintf(fmt, Qpixel ? (real)i : x);
	      }
	      printf("\n\n");
	      nlcount += 2;
            }
            if (!newline && ylabel) {            /* print first column of Y coord */
	      myprintf(fmt, Qpixel ? (real) j : y);
	      printf(" ");
            }
            for (i=0; i<nxpos; i++) {			/* loop over all columns */
  	        if (Qpair && nxpos > 1) i = l;
                f = CubeValue(iptr,ix[i],iy[j],iz[k]);
                if (newline) {
		  x = Xmin(iptr) + (ix[i]-Xref(iptr)) * Dx(iptr);
                    if (xlabel) myprintf(fmt,Qpixel ? (real)i : x);
                    if (ylabel) myprintf(fmt,Qpixel ? (real)j : y);
                    if (zlabel) myprintf(fmt,Qpixel ? (real)k : z);
                } 
                printf (fmt,f*scale_factor);
                if (newline) {
                    printf("\n"); nlcount++;
                } else {
                    printf (" ");
		}
		if (Qpair) {
		  i=nxpos; /* make sure it doesn't loop here */
		  l++;     /* but increase the global loop counter */
		}
            } /* i */
            if (newline1) {
	      printf("\n");
	      nlcount++;
	    }
	    if (Qpair) j1 = nypos; /* make sure it doesn't loop here */
        } /* j1 */
        if (newline2) {
	  printf("\n\n");
	  nlcount += 2;
	}
	if (Qpair) k = nzpos; /* make sure it doesn't loop here */
    } /* k */
    if (nlcount == 0) printf("\n");
    strclose(instr);
}

int ini_array(
	  string key,             /* keyword */
	  int *dat,               /* array */
	  int ndat,               /* length of array */
	  int offset)             /* offset applied to index array */
{
    int i, n;

    if (ndat <= 0) return ndat;
    n = hasvalue(key) ? nemoinpi(getparam(key),dat,ndat) : 0;
    if (n > 0) {
        for (i=0; i<n; i++) {
            dat[i] -= offset;
            if (dat[i] < 0 || dat[i] >= ndat)
	      error("Illegal boundary in %s",key);
        }
    } else if (n==0) {     /* select all axis elements */
        for (i=0; i<ndat; i++)
	  dat[i] = i;
        n = ndat;
    } else
      error("Error %d parsing %s=%s",n,key,getparam(key));
    return n;
}

void myprintf(string fmt,real v)
{
    printf(fmt,v);
    printf(" ");
}
